sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1902,249 @@ def transform(self, X):
         """
         return _transform_selected(X, self._transform,
                                    self.categorical_features, copy=True)
+
+
+def _get_count_dict_3():
+    """Gets the outer count dictionary."""
+
+    # get_count_dict() is an optimization on key-value access
+    # where we have multi-layer dicts for accessing dict[X][y] instead of
+    # using dict[(X, y)], because using dict[(X, y)] is inefficient
+    # Using external helper methods is also required to avoid a picking
+    # error involving lambda expressions in defaultdict
+    return defaultdict(_get_count_dict_2)
+
+
+def _get_count_dict_2():
+    """Gets the inner count dictionary."""
+    return defaultdict(_get_count_dict_1)
+
+
+def _get_count_dict_1():
+    """Gets the inner count dictionary."""
+    return defaultdict(_get_count_dict_0)
+
+
+def _get_count_dict_0():
+    """Gets the innermost count dictionary."""
+    return defaultdict(int)
+
+
+class CountFeaturizer(BaseEstimator, TransformerMixin):
+    """Adds a feature representing each feature value's count in training
+
+    Specifically, for each data point 'X_i' in the dataset 'X', it will add in
+    a new set of columns 'count_X_i' to the end of 'X_i' where 'count_X_i' is the
+    number of occurences of 'X_i' in the dataset 'X' given the equality indicator
+    'inclusion'.
+
+    If a 'y' argument is given during the fit step, then the
+    count in the transform step will be conditional on the 'y'.
+    The number of columns added will be the number of different values 'y' can
+    take on.
+
+    This preprocessing step is useful when the number of occurences
+    of a particular piece of data is helpful in computing the prediction.
+
+    Parameters
+    ----------
+    inclusion : 'all', 'each', list, or numpy.ndarray
+        The inclusion criteria for counting
+
+        - 'all' (default) : Every feature given is counted
+        - 'each' : Each feature will have its own set of counts
+        - 1D list of indices : Only the given list of features is
```

Comment:
Changed to

        - 1D list of indices : Only the given list of features is
        concatenated and counted
        - 2D list of indices : The given list of lists of features is
        concatenated and counted, but each list in the list of lists
        has its own set of counts